interop-cats/shared/src/main/scala/zio/interop/cats.scala

```diff
@@ -35,348 +38,443 @@ abstract class CatsEffectPlatform
+  private[this] val monadErrorInstance0: MonadError[Task, Throwable] =
+    new ZioMonadError
 }
 
-private class CatsDefer[R, E] extends Defer[ZIO[R, E, *]] {
-  def defer[A](fa: => ZIO[R, E, A]): ZIO[R, E, A] = ZIO.effectSuspendTotal(fa)
+private class ZioDefer[R, E] extends Defer[ZIO[R, E, *]] {
+  type F[A] = ZIO[R, E, A]
+
+  override final def defer[A](fa: => F[A]): F[A] =
+    ZIO.effectSuspendTotal(fa)
 }
 
-private class CatsConcurrentEffect[R](rts: Runtime[R])
-    extends CatsConcurrent[R]
-    with effect.ConcurrentEffect[RIO[R, *]] {
-
-  override final def runAsync[A](fa: RIO[R, A])(
-    cb: Either[Throwable, A] => effect.IO[Unit]
-  ): effect.SyncIO[Unit] =
-    effect.SyncIO {
-      rts.unsafeRunAsync(fa.run) { exit =>
-        cb(exit.flatMap(identity).toEither).unsafeRunAsync(_ => ())
-      }
-    }
+private class ZioConcurrent[R, E] extends ZioMonadError[R, E] with GenConcurrent[ZIO[R, E, *], E] {
 
-  override final def runCancelable[A](fa: RIO[R, A])(
-    cb: Either[Throwable, A] => effect.IO[Unit]
-  ): effect.SyncIO[effect.CancelToken[RIO[R, *]]] =
-    effect.SyncIO {
-      rts.unsafeRun {
-        ZIO.descriptor
-          .bracketExit(
-            (descriptor, exit: Exit[Throwable, A]) =>
-              ZIO.effectTotal {
-                exit match {
-                  case Exit.Failure(cause) if !cause.interruptors.forall(_ == descriptor.id) => ()
-                  case _ =>
-                    effect.IO.suspend(cb(exit.toEither)).unsafeRunAsync(_ => ())
-                }
-              },
-            _ => fa
-          )
-          .interruptible
-          .forkDaemon
-          .map(_.interrupt.unit)
-      }
-    }
+  private def toPoll(restore: ZIO.InterruptStatusRestore) = new Poll[ZIO[R, E, *]] {
+    override def apply[T](fa: ZIO[R, E, T]): ZIO[R, E, T] = restore(fa)
+  }
+
+  private def toFiber[A](fiber: Fiber[E, A]) = new effect.Fiber[F, E, A] {
+    override final val cancel: F[Unit]           = fiber.interrupt.unit
+    override final val join: F[Outcome[F, E, A]] = fiber.await.map(toOutcome)
+  }
+
+  private def fiberFailure(error: E) =
+    FiberFailure(Cause.fail(error))
+
+  override def ref[A](a: A): F[effect.Ref[F, A]] =
+    ZRef.make(a).map(new ZioRef(_))
+
+  override def deferred[A]: F[Deferred[F, A]] =
+    Promise.make[E, A].map(new ZioDeferred(_))
 
-  override final def toIO[A](fa: RIO[R, A]): effect.IO[A] =
-    effect.ConcurrentEffect.toIOFromRunCancelable(fa)(this)
+  override final def start[A](fa: F[A]): F[effect.Fiber[F, E, A]] =
+    fa.fork.map(toFiber)
+
+  override def never[A]: F[A] =
+    ZIO.never
+
+  override final def cede: F[Unit] =
+    ZIO.yieldNow
+
+  override final def forceR[A, B](fa: F[A])(fb: F[B]): F[B] =
+    fa.either *> fb
+
+  override final def uncancelable[A](body: Poll[F] => F[A]): F[A] =
+    ZIO.uninterruptibleMask(body.compose(toPoll))
+
+  override final val canceled: F[Unit] =
+    ZIO.halt(Cause.interrupt(Fiber.Id.None))
```

Comment:
There is a significant difference between the ZIO and Cats Effect 3 interruption model here that I am not sure how we handle. In Cats Effect 3 self-interruption is ignored in an uninterruptible region, so we would have to check the interrupt status here and return `ZIO.interrupt` only if we are in an interruptible region. However, under the Cats Effect 3 model self interruption in an uninterruptible region is still observed at the end of the region which doesn't really make sense in the ZIO model because we treat an interruption as a value that can be handled with operators such as `run`.

Reply:
I must admit I don't quite understand this.

> which doesn't really make sense in the ZIO model because we treat an interruption as a value that can be handled with operators such as run.

Why can't the same be done in CE3? It also has `Outcome.Canceled`

Reply:
It seems like there is enough flexibility in the laws that this isn't an issue but consider:

```scala
ZIO.uninterruptibleMask(_ => ZIO.interrupt *> ZIO.effectTotal(println("I'm here"))
```

Versus the equivalent in Cats Effect 3:

```scala
F.uncancelable(_ => F.canceled *> F.delay(println("I'm here"))
```

In ZIO, "I'm here" will never be printed because while the uninterruptible region prevents the fiber from being interrupted externally `ZIO.interrupt` causes the fiber to die with an interruption cause, which in turn could be recovered from using operators like `run`.

In Cats Effect 3, in contrast, what they call "self-cancelation" is suppressed in an uncancelable region, so "I'm here" will be printed, but then the fiber would be subject to cancelation as soon as it completed the uncancelable region.

You can see this in the difference of the signature of `interrupt`, which returns a `UIO[Nothing`, and `canceled`, which must return a `F[Unit]` because it may actually return successfully if it is in an uninterruptible region.

It seems like we are satisfying the laws with the latest changes so this may not be an issue for this PR in particular but it is a pretty significant difference in the interruption model with a lot of other implications.

Reply:
Hmm ok I see, thanks for the explanation - I hope this difference wouldn't be a problem in practice.